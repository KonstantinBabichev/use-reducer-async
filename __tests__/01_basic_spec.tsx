import React, {
  StrictMode,
  Reducer,
  Dispatch,
} from 'react';

import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';

import { useReducerAsync } from 'use-reducer-async';

describe('basic spec', () => {
  afterEach(cleanup);

  it('sleep', async () => {
    type State = {
      sleeping: boolean;
    };
    const initialState: State = {
      sleeping: false,
    };
    type Action =
      | { type: 'START_SLEEP' }
      | { type: 'END_SLEEP' };
    const reducer: Reducer<State, Action> = (state, action) => {
      switch (action.type) {
        case 'START_SLEEP': return { ...state, sleeping: true };
        case 'END_SLEEP': return { ...state, sleeping: false };
        default: throw new Error('no such action type');
      }
    };
    type AsyncAction = { type: 'SLEEP'; ms: number };
    const asyncActions = {
      SLEEP: (dispatch: Dispatch<Action>) => async (action: AsyncAction) => {
        dispatch({ type: 'START_SLEEP' });
        await new Promise(r => setTimeout(r, action.ms));
        dispatch({ type: 'END_SLEEP' });
      },
    };
    const Component = () => {
      const [state, dispatch] = useReducerAsync<
        Reducer<State, Action>,
        AsyncAction,
        AsyncAction | Action
      >(
        reducer,
        initialState,
        asyncActions,
      );
      return (
        <div>
          <span>{state.sleeping ? 'Sleeping' : 'Idle'}</span>
          <button type="button" onClick={() => dispatch({ type: 'SLEEP', ms: 100 })}>Click</button>
        </div>
      );
    };
    const App = () => (
      <StrictMode>
        <Component />
      </StrictMode>
    );
    const { getAllByText, container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(getAllByText('Click')[0]);
    expect(container).toMatchSnapshot();
    await waitForElement(() => getAllByText('Idle'));
    expect(container).toMatchSnapshot();
  });
});
