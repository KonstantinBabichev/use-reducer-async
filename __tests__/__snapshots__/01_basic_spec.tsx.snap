// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic spec sleep 1`] = `
<div>
  <div>
    <span>
      Idle
    </span>
    <button
      type="button"
    >
      Click
    </button>
  </div>
</div>
`;

exports[`basic spec sleep 2`] = `
<div>
  <div>
    <span>
      Sleeping
    </span>
    <button
      type="button"
    >
      Click
    </button>
  </div>
</div>
`;

exports[`basic spec sleep 3`] = `
<div>
  <div>
    <span>
      Idle
    </span>
    <button
      type="button"
    >
      Click
    </button>
  </div>
</div>
`;
